packages/@aws-cdk/aws-codepipeline-api/lib/action.ts

```diff
@@ -270,16 +222,46 @@ export abstract class Action extends cdk.Construct {
     );
   }
 
-  protected addOutputArtifact(name: string = this.stage._internal._generateOutputArtifactName(this)): Artifact {
-    const artifact = new Artifact(this, name);
+  protected addOutputArtifact(name: string): Artifact {
+    const artifact = new Artifact(name);
     this._actionOutputArtifacts.push(artifact);
     return artifact;
   }
 
-  protected addInputArtifact(artifact: Artifact = this.stage._internal._findInputArtifact(this)): Action {
+  protected addInputArtifact(artifact: Artifact): Action {
     this._actionInputArtifacts.push(artifact);
     return this;
   }
+
+  protected getParent(methodName: string): cdk.Construct {
+    if (this.parent) {
+      return this.parent;
+    } else {
+      throw new Error(`Cannot call method '${methodName}' until the Action has been added to a Pipeline with addAction`);
+    }
+  }
+
+  protected abstract bind(pipeline: IPipeline, parent: cdk.Construct): void;
+
+  // ignore unused private method (it's actually used in Stage)
+  // @ts-ignore
+  private _attachActionToPipeline(pipeline: IPipeline, stage: IStage, parent: cdk.Construct): void {
```

Comment:
> rename parent to scope

Done.

Reply:
> Wouldn't it be sufficient to just associate the action to the stage (and then use stage.pipeline to reference the pipeline). In general, try to reduce the amount of state you synchronize across the layers.

It would not. We also need the Pipeline there, hence the 2 parameters (the Pipeline is the only Construct in the hierarchy, so we need it for anchoring).

Reply:
I understand we need the pipeline, but you should be able to access the pipeline from the stage, no?

Reply:
Ah. I misunderstood what you wanted here, sorry about that.

Sure, changed.